enhancements/rhcos/split-rhcos-into-layers.md

```diff
@@ -0,0 +1,458 @@
+---
+title: split-rhcos-into-layers
+authors:
+  - "@jlebon"
+  - "@cverna"
+  - "@travier"
+reviewers:
+  - "@patrickdillon, for installer impact"
+  - "@rphillips, for node impact"
+  - "@joepvd, for ART impact"
+  - "@yuqi-zhang, for MCO impact"
+  - "@LorbusChris, for OKD impact"
+  - "@zaneb, for agent installer impact"
+  - "@sdodson, for overall architecture"
+  - "@cgwalters, for overall architecture"
+  - "@jupierce, for overall architecture"
+approvers:
+  - "@mrunalp"
+api-approvers:
+  - None
+creation-date: 2024-06-07
+last-updated: 2024-06-07
+tracking-link:
+  - https://issues.redhat.com/browse/OCPSTRAT-1190
+see-also:
+  - "/enhancements/ocp-coreos-layering/ocp-coreos-layering.md"
+---
+
+# Split RHCOS into layers
+
+## Summary
+
+This enhancement describes improvements to the way RHEL CoreOS (RHCOS) is built
+so that it will better align with image mode for RHEL, all while also providing
+benefits on the OpenShift side. Currently, RHCOS is built as a single layer that
+includes both RHEL and OCP content. This enhancement proposes splitting it into
+three layers. Going from bottom to top:
+1. the bootc layer; i.e. the base rhel-bootc image shared with image mode for RHEL (RHEL-versioned)
+2. the CoreOS layer; i.e. coreos-installer, ignition, afterburn, scripts, etc... (RHEL-versioned)
+3. the node layer; i.e. oc, kubelet, cri-o, etc... (OCP-versioned)
+
+The terms "bootc layer", "CoreOS layer", and "node layer" will be used
+throughout this enhancement to refer to these.
+
+The details of this enhancement focus on doing the first split: creating the node
+layer as distinct from the CoreOS layer (which will not yet be rebased on top of
+a bootc layer). The two changes involved which most affect OCP are:
+1. bootimages will no longer contain OCP components (e.g. kubelet, cri-o, etc...)
+2. the `rhel-coreos` payload image will be built in Prow/Konflux (as any other)
+
+This enhancement is based on initial sketches by @cgwalters in:
+https://github.com/openshift/os/issues/799
+
+## Motivation
+
+Image mode for RHEL provides an opportunity to rethink how RHCOS is built
+and allow us to share a lot more with other image mode variants. A key benefit
+on the RHEL side is that we would have only one stream of RHCOS per RHEL release,
+rather than one per OpenShift release. This greatly reduces the workload on the
+CoreOS team. Another benefit is easier integration in the CI processes of
+rhel-bootc and centos-bootc, as well as better shared documentation.
+
+On the OCP side, the final node image would now be built as a layered
+container image, just like most other OpenShift components are built. This should
+allow simplifying CI and ART tooling related to RHCOS, much of which is bespoke.
+
+For example, ART would no longer have to sync RHCOS images to CI; CI could
```

Comment:
Thanks, that's a detail I didn't know but it makes sense. So overall images in the CI imagestreams may come from either CI _or_ ART nightlies depending on how fresh the repo is?

I'm trying to verify my understanding of this by comparing the image digests in e.g. registry.ci.openshift.org/ocp/release:4.17.0-0.ci-2024-09-24-105044 vs registry.ci.openshift.org/ocp/release:4.17.0-0.nightly-2024-09-23-211417; I would've expected some of them to be the same there. The CI payloads are manifest listed but not the nightly ART one, but even comparing against the amd64 digests in the manifest lists against the nightly ones, I'm not getting any component images that match other than the RHCOS ones.

Hmm, and based on [this other comment](https://github.com/openshift/enhancements/pull/1637/files#r1755881465), are you suggesting not having CI builds promote at all and keep relying on just ART or both? I'd like the CoreOS images to be at parity with the other component images. E.g. merges to openshift/os would ideally trigger a CI build and promotion.

Reply:
> I'm not getting any component images that match other than the RHCOS ones.

ART pushes RHCOS and _base_ images to the _integration_ imagestream. The latter serve as base images for other component builds that wind up in the release payload (e.g. https://github.com/openshift/oc/blob/3692450b96d57ae3870e5f693e833a0701d0e2b0/images/cli/Dockerfile.rhel#L6 ) . The base images themselves do not show up in the payload.

> not having CI builds promote at all and keep relying on just ART or both?

We could:
1. Leave things as they are today (ART pushes RHCOS imagestreams to the CI integration imagestream).
2. Have only CI promote rhcos into the CI integration imagestream.
3. Have both push.

The downside of #1 is that is not fast. It may take hours for an RHCOS merge to reach the CI integration imagestream by way of ART.

The downside of #2 is that the production integration imagestream can diverge, for significant periods of time, from the CI integration imagestream. CI builds & promotion generally only happen on merge in the component's repository. A merge in openshift/hyperkube with important updates to the kubelet would not cause #2 to rebuild RHCOS. This can lead to difficult to diagnose disparities (e.g. nightlies passing tests but CI payloads failing).

The downside of #3 is that it is obscure and provenance of the RHCOS image in the CI integration stream can flip back and forth between ART and CI builds. ART should also put in a check to ensure that it doesn't update the imagestream with an image that is older than one CI has already promoted.

Reply:
> ART pushes RHCOS and base images to the integration imagestream. The latter serve as base images for other component builds that wind up in the release payload

Gotcha, I follow now. ART does pushes to e.g. registry.ci.openshift.org/ocp/4.18:base-rhel9. This then triggers a rebuild _in the CI cluster_ of all the images based on that base image.

> We could:
>  1. Leave things as they are today (ART pushes RHCOS imagestreams to the CI integration imagestream).
>  2. Have only CI promote rhcos into the CI integration imagestream.
>  3. Have both push.

Before digging into that, let me clarify again the images in play here:
- there's the CoreOS base image containing the pure RHEL/CentOS Stream content (I've been calling this the "rhel-coreos-base" image)
- there's the OCP node image which builds on top of the rhel-coreos-base image and adds all the OCP-specific bits (which confusingly will still be called "rhel-coreos" in the release payload but ideally we'd rename to e.g. "node")

The rhel-coreos-base image will keep being built in the internal CoreOS pipeline (currently Jenkins, in the future Konflux).

What I'm trying to see is if it makes sense to have the OCP node image be setup in a similar way to the other OCP components. So e.g. taking openshift/oc as an example, which is built on top of the base-rhel9 base image. It has both a CI build which gets promoted to the CI imagestream and a prod ART build which gets pushed to the nightly imagestream (right?).

So then, two approaches I see for handling RHCOS are:
1. Only have ART build the node image. It then syncs it to the CI imagestream.
2. Have an ART build of the node image _and_ a CI build of the node image. The CI build is triggered by changes in openshift/os, but also by changes to its base image (rhel-coreos-base), which ART or the CoreOS pipeline itself would push to the CI imagestream.

(1) is closer I think to what we're currently doing. (2) is closer to how other OCP components are set up and is what I'm currently proposing.

As you correctly noted, one key aspect of the node image is that it consumes content from other OCP components as RPMs built by ART. As mentioned in Phase 3, ART will have to adapt its logic to detect if OCP RPMs in the node image are stale and we need to rebuild. In (2), it could be the one that also retriggers the CI build at the same time it triggers its own build to ensure CI doesn't go stale but without muddling provenance.

Reply:
Chatted with Justin about this. There was agreement to keep things simple to start and just have ART also push to the CI imagestream.